drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+
+	*config->regs.ctr = v;
+
+	switch (dev_config.bits.speed) {
+	case I2C_SPEED_STANDARD:
+		i2c_esp32_configure_frequency(config, 100000);
+		break;
+	case I2C_SPEED_FAST:
+		i2c_esp32_configure_frequency(config, 400000);
+		break;
+	default:
+		r = -EINVAL;
+		goto out;
+	}
+
+	/* Use FIFO to transmit data */
+	*config->regs.fifo_cfg &= ~I2C_NONFIFO_EN;
+
+	v = CONFIG_I2C_ESP32_TIMEOUT & I2C_TIME_OUT_REG;
+	*config->regs.timeout = v << I2C_TIME_OUT_REG_S;
+
+	/* Enable interrupt types handled by the ISR */
+	*config->regs.intr |= I2C_ACK_ERR_INT_ENA_M |
+			      I2C_TIME_OUT_INT_ENA_M |
+			      I2C_TRANS_COMPLETE_INT_ENA_M |
+			      I2C_ARBITRATION_LOST_INT_ENA_M;
+
+	/* Activate the device IRQ */
+	irq_enable(config->irq.line);
+
+out:
+	irq_unlock(key);
+
+	return r;
+}
+
+static inline void i2c_esp32_ack_intrs(const struct i2c_esp32_config *config)
+{
+	/* Acknowledge all interrupts */
+	*config->regs.intr_clr = ~0;
+}
+
+
+static inline void i2c_esp32_reset_fifo(const struct i2c_esp32_config *config)
+{
+	/* Writing 1 and then 0 to these bits will reset the I2C fifo */
+	*config->regs.fifo_cfg |= (I2C_TX_FIFO_RST | I2C_RX_FIFO_RST);
+	*config->regs.fifo_cfg &= ~(I2C_TX_FIFO_RST | I2C_RX_FIFO_RST);
+}
+
+static inline void i2c_esp32_prepare(const struct i2c_esp32_config *config)
+{
+	i2c_esp32_ack_intrs(config);
+	i2c_esp32_reset_fifo(config);
+}
+
+static int i2c_esp32_spin_yield(int *counter)
+{
+	*counter = *counter + 1;
+
+	if (*counter > I2C_ESP32_TIMEOUT) {
+		return -ETIMEDOUT;
+	}
+
+	if (*counter > I2C_ESP32_SPIN_THRESHOLD) {
+		k_yield();
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_transmit_and_wait(struct device *dev,
+				       volatile struct i2c_esp32_cmd *wait_cmd)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	u32_t status;
+	int counter = 0;
+	int r;
+
+	i2c_esp32_ack_intrs(config);
+
+	/* Start transmission and wait for the ISR to give the semaphore */
+	*config->regs.ctr |= I2C_TRANS_START;
+	if (k_sem_take(&data->fifo_sem, I2C_ESP32_TIMEOUT_MS) < 0) {
+		return -ETIMEDOUT;
+	}
+
+	/* Check for error conditions before waiting for the bus to finish */
+	status = *config->regs.intr_raw;
+	if (status & (I2C_ARBITRATION_LOST_INT_RAW | I2C_ACK_ERR_INT_RAW)) {
+		return -EIO;
+	}
+	if (status & I2C_TIME_OUT_INT_RAW) {
+		return -ETIMEDOUT;
+	}
+
+	if (wait_cmd) {
+		while (!wait_cmd->done) {
+			r = i2c_esp32_spin_yield(&counter);
+			if (r < 0) {
+				return r;
+			}
+		}
+	}
+
+	/* Wait for I2C bus to finish its business */
+	while (*config->regs.status & I2C_BUS_BUSY) {
+		r = i2c_esp32_spin_yield(&counter);
+		if (r < 0) {
+			return r;
+		}
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_read_msg(struct device *dev, u16_t addr,
+			      struct i2c_msg msg)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	volatile struct i2c_esp32_cmd *cmd_base = (struct i2c_esp32_cmd *)config->regs.cmd_base;
```

Comment:
too long line it seems

Reply:
Useful trick: make the cast a void*.  Only six characters for the type and guaranteed to assign cleanly to any pointer lvalue.  In a declaration like this it's where the type is already present it hurts nothing and improves readability.

Reply:
I'll just change the config struct and make the cast where it's initialized.